torch_geometric/explain/algorithm/gnn_explainer.py

```diff
@@ -0,0 +1,318 @@
+from math import sqrt
+from typing import Optional, Tuple, Union
+
+import torch
+from torch import Tensor
+from torch.nn.parameter import Parameter
+
+from torch_geometric.explain.algorithm.utils import clear_masks, set_masks
+from torch_geometric.explain.config import (
+    ExplainerConfig,
+    MaskType,
+    ModelConfig,
+    ModelMode,
+    ModelReturnType,
+    ModelTaskLevel,
+)
+from torch_geometric.explain.explanations import Explanation
+
+from .base import ExplainerAlgorithm
+
+EPS = 1e-15
+
+
+class GNNExplainer(ExplainerAlgorithm):
+    r"""The GNN-Explainer model from the `"GNNExplainer: Generating
+    Explanations for Graph Neural Networks"
+    <https://arxiv.org/abs/1903.03894>`_ paper for identifying compact subgraph
+    structures and small subsets node features that play a crucial role in a
+    GNN's node-predictions.
+
+    .. note::
+
+        For an example of using GNN-Explainer, see `examples/gnn_explainer.py
+        <https://github.com/pyg-team/pytorch_geometric/blob/master/examples/
+        gnn_explainer.py>`_.
+
+    Args:
+        epochs (int, optional): The number of epochs to train.
+            (default: :obj:`100`)
+        lr (float, optional): The learning rate to apply.
+            (default: :obj:`0.01`)
+        shared_mask (bool, optional): Whether to share the mask for node
+            features. When set to :obj:`True`, the node feature mask will be
+            shared across all nodes. (default: :obj:`True`). Only used when the
+            node mask type is :obj:`MaskType.attributes`.
+        **kwargs (optional): Additional hyper-parameters to override default
+            settings in
+            :attr:`~torch_geometric.explain.algorithm.GNNExplainer.coeffs`.
+    """
+
+    coeffs = {
+        'edge_size': 0.005,
+        'edge_reduction': 'sum',
+        'node_feat_size': 1.0,
+        'node_feat_reduction': 'mean',
+        'edge_ent': 1.0,
+        'node_feat_ent': 0.1,
+    }
+
+    def __init__(
+        self,
+        epochs: int = 100,
+        lr: float = 0.01,
+        shared_mask: bool = True,
+        **kwargs,
+    ):
+        super().__init__()
+        self.epochs = epochs
+        self.lr = lr
+        self.coeffs.update(kwargs)
+        self.shared_mask = shared_mask
+        self.node_mask = None
+        self.edge_mask = None
+
+    def forward(self, model: torch.nn.Module, x: Tensor, edge_index: Tensor,
+                explainer_config: ExplainerConfig, model_config: ModelConfig,
+                target: Tensor, target_index: Optional[Union[int,
+                                                             Tensor]] = None,
+                index: Optional[Union[int, Tensor]] = None,
```

Comment:
```suggestion
                index: Optional[int] = None,
```
At the moment the `loss` function only supports training for one node.